File: fuse/python/src/fusepy/objects.cc
#include "objects.h"
#include "addresses.h"

using namespace fuse;

auto toPython(const AnnotatedBinary& binary) -> PyObject*
{
    PyObject* pyObjectList = PyList_New(binary.objectCount());
    for (size_t i = 0; i < binary.objectCount(); ++i)
    {
        auto const& object = *binary.object(i);
        PyObject* pyObject = toPython(object);

        auto const offset = static_cast<const PackedObject&>(object).offset();
        PyObject* pyBinary = PyByteArray_FromStringAndSize(
            reinterpret_cast<const char*>(binary.binary().data() + offset), object.size());
        PyDict_SetItemString(pyObject, "binary", pyBinary);

        PyList_SetItem(pyObjectList, i, pyObject);
    }
    return pyObjectList;
}

auto toPython(const Object::Section& section) -> PyObject*
{
    PyObject* pySection = PyDict_New();
    PyObject* pyOffset = Py_BuildValue("K", static_cast<unsigned long long>(section.offset));
    PyDict_SetItemString(pySection, "offset", pyOffset);
    PyObject* pyActualOffset =
        Py_BuildValue("K", static_cast<unsigned long long>(section.realOffset));
    PyDict_SetItemString(pySection, "actual_offset", pyActualOffset);
    PyObject* pySize = Py_BuildValue("K", static_cast<unsigned long long>(section.size));
    PyDict_SetItemString(pySection, "size", pySize);
    return pySection;
}

auto toPython(const Object& object) -> PyObject*
{
    PyObject* pyObject = PyDict_New();

    PyObject* pyPath = PyUnicode_FromStringAndSize(object.path().toString().c_str(),
                                                   object.path().toString().length());
    PyDict_SetItemString(pyObject, "path", pyPath);

    if (object.hasFixedOffset())
    {
        PyDict_SetItemString(
            pyObject, "fixed_offset",
            Py_BuildValue("K", static_cast<unsigned long long>(object.fixedOffset())));
    }

    if (object.alignment() != 1)
    {
        PyDict_SetItemString(
            pyObject, "alignment",
            Py_BuildValue("K", static_cast<unsigned long long>(object.alignment())));
    }

    PyObject* pyActualSize = Py_BuildValue("K", static_cast<unsigned long long>(object.realSize()));
    PyDict_SetItemString(pyObject, "actual_size", pyActualSize);

    PyObject* pyRefList = PyList_New(object.unresolvedReferenceCount());
    for (size_t i = 0; i < object.unresolvedReferenceCount(); ++i)
    {
        PyObject* pyRef = toPython(object.unresolvedReference(i));
        PyList_SetItem(pyRefList, i, pyRef);
    }
    PyDict_SetItemString(pyObject, "unresolved", pyRefList);

    PyObject* pySectionList = PyList_New(object.sectionCount());
    for (size_t i = 0; i < object.sectionCount(); ++i)
    {
        PyObject* pySection = toPython(object.section(i));
        PyList_SetItem(pySectionList, i, pySection);
    }
    PyDict_SetItemString(pyObject, "sections", pySectionList);

    return pyObject;
}

auto toPython(const UnresolvedReference& ref) -> PyObject*
{
    PyObject* pyRef = PyDict_New();
    PyObject* pyOffset = Py_BuildValue("K", static_cast<unsigned long long>(ref.relativeOffset()));
    PyDict_SetItemString(pyRef, "offset", pyOffset);
    PyObject* pyPath = PyUnicode_FromStringAndSize(ref.referencedPath().toString().c_str(),
                                                   ref.referencedPath().toString().length());
    PyDict_SetItemString(pyRef, "path", pyPath);

    PyObject* pyLayout = PyAddressLayout_New(ref.addressLayout().copy());
    PyDict_SetItemString(pyRef, "layout", pyLayout);

    return pyRef;
}

File: fuse/python/fusepy/objects.py
#from fusepy._fusepy import 
class UnresolvedReference:
    def __init__(self, offset, path, layout):
        self.offset = offset
        self.path = path
        self.layout = layout

    def __repr__(self):
        return f'UnresolvedReference({self.offset}, {self.path}, {self.layout})'

    def to_dict(self):
        return {
            'offset': self.offset,
            'path': self.path,
            'layout': self.layout.to_dict()
        }

class ObjectSection:
    def __init__(self, offset, actual_offset, size):
        self.offset = offset
        self.actual_offset = actual_offset
        self.size = size

    def __repr__(self):
        return f'ObjectSection({self.offset}, {self.actual_offset}, {self.size})'

    def to_dict(self):
        return {
            'offset': self.offset,
            'actual_offset': self.actual_offset,
            'size': self.size,
        }

class PackedObject:
    def __init__(self, path, offset, size, actual_size, sections, unresolved):
        self.path = path
        self.offset = offset
        self.size = size
        self.actual_size = actual_size
        self.sections = sections
        self.unresolved = unresolved

    def to_dict(self):
        sections = []
        for section in self.sections:
            sections.append(section.to_dict())
        unresolved = []
        for ref in self.unresolved:
            unresolved.append(ref.to_dict())
        return {
            'path': self.path,
            'offset': self.offset,
            'size': self.size,
            'actual_size': self.actual_size,
            'sections': sections,
            'unresolved': unresolved,
        }

def _numerus(count, singular, plural):
    if count == 1:
        return singular
    else:
        return plural

class PackedObjects:
    def __init__(self, binary, objects):
        self.binary = binary
        self.objects = objects

    def __str__(self):
        return f'{len(self.objects)} {_numerus(len(self.objects), "object", "objects")} packed into {len(self.binary)} bytes'

    def to_dict(self):
        objects = []
        for obj in self.objects:
            objects.append(obj.to_dict())
        return {
            'binary': self.binary,
            'objects': objects,
        }